### PR TITLE
add image_fusion to plantcv

### PR DESCRIPTION
**Pull request by DannieSheng**
**Describe your changes**
Add `image_fusion` to PlantCV. 
(The 1st application would be fuse nir and RGB image, create a data cube that belongs to PlantCV.Spectraldata class)

**Type of update**
Is this a:
* New feature

### CODE REVIEW

File: plantcv/plantcv/__init__.py [hunk 1]
```diff
@@ -94,8 +95,8 @@
 
 __all__ = ['fatal_error', 'Params', 'Outputs', 'Spectral_data', 'deprecation_warning', 'print_image', 'plot_image',
            'color_palette', 'apply_mask', 'gaussian_blur', 'transform', 'hyperspectral', 'readimage', 'readbayer',
-           'laplace_filter', 'sobel_filter', 'scharr_filter', 'hist_equalization', 'erode', 'image_add', 
-           'image_subtract', 'dilate', 'watershed', 'rectangle_mask', 'rgb2gray_hsv', 'rgb2gray_lab', 'rgb2gray_cmyk',
```
Comment (nfahlgren) [inline plantcv/plantcv/__init__.py:98]: `rgb2gray_cmyk` was accidentally removed from `__all__`

File: plantcv/plantcv/image_fusion.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+import os
+import numpy as np
+from plantcv.plantcv import fatal_error
+from plantcv.plantcv import Spectral_data
+from plantcv.plantcv import params
+from plantcv.plantcv.hyperspectral.read_data import _make_pseudo_rgb
+from plantcv.plantcv._debug import _debug
+
+
+def image_fusion(img1, img2, wvs1, wvs2, array_type = None, filename = None):
```
Comment (nfahlgren) [inline plantcv/plantcv/image_fusion.py:12]:
```suggestion
def image_fusion(img1, img2, wvs1, wvs2, array_type=None, filename=None):
```

No spaces between keywords and default values

File: plantcv/plantcv/image_fusion.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    :param img2: np.ndarray
+    :param wvs1: list
+    :param wvs2: list
+    :param array_type: str
+    :param filename: str
+    :return: fused_array (a Spectral_data instance)
+    """
+
+    if len(img1.shape) == 2:
+        img1 = np.expand_dims(img1,axis=2)
```
Comment (nfahlgren) [inline plantcv/plantcv/image_fusion.py:31]:
```suggestion
        img1 = np.expand_dims(img1, axis=2)
```

Add one whitespace between arguments

File: plantcv/plantcv/image_fusion.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    r1, c1, b1 = img1.shape
+
+    if len(img2.shape) == 2:
+        img2 = np.expand_dims(img2,axis=2)
```
Comment (nfahlgren) [inline plantcv/plantcv/image_fusion.py:35]:
```suggestion
        img2 = np.expand_dims(img2, axis=2)
```

Add one whitespace between arguments

File: plantcv/plantcv/image_fusion.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    r2, c2, b2 = img2.shape
+    if (r1,c1) != (r2,c2):
```
Comment (nfahlgren) [inline plantcv/plantcv/image_fusion.py:37]:
```suggestion
    if (r1, c1) != (r2, c2):
```

Add one whitespace between elements in lists and tuples

File: plantcv/plantcv/image_fusion.py [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        fatal_error("Input images should have the same image size")
+
+    array_data  = np.concatenate((img1, img2), axis=2)
```
Comment (nfahlgren) [inline plantcv/plantcv/image_fusion.py:40]:
```suggestion
    array_data = np.concatenate((img1, img2), axis=2)
```

Remove extra spaces

File: plantcv/plantcv/image_fusion.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    wavelengths = np.array(wvs1 + wvs2)
+    ind = np.argsort(wavelengths)
+    wavelengths = wavelengths[ind]
+
+    wavelength_dict = dict()
+    for (idx, wv) in enumerate(wavelengths):
+        wavelength_dict[wv] = float(idx)
+
+    # sort array_data based on wavelengths
+    array_data = array_data[:,:,ind]
```
Comment (nfahlgren) [inline plantcv/plantcv/image_fusion.py:52]:
```suggestion
    array_data = array_data[:, :, ind]
```

Add missing spaces after commas

File: plantcv/plantcv/image_fusion.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+
+    fused_array = Spectral_data(array_data=array_data,
+                                   max_wavelength=float(max(wavelengths)),
+                                   min_wavelength=float(min(wavelengths)),
+                                   max_value=max_pixel, min_value=min_pixel,
+                                   d_type=d_type,
+                                   wavelength_dict=wavelength_dict, samples=int(r * c),
+                                   lines=int(b), interleave="bil",
+                                   wavelength_units="nm", array_type=array_type,
+                                   pseudo_rgb=None, filename=filename, default_bands=None)
```
Comment (nfahlgren) [inline plantcv/plantcv/image_fusion.py:70]: Indented lines should line up with the first argument (`array_data`)

File: plantcv/plantcv/image_fusion.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+
+    # Make pseudo-rgb image and replace it inside the class instance object
+    pseudo_rgb = _make_pseudo_rgb(fused_array)
+    fused_array.pseudo_rgb = pseudo_rgb
+
+    _debug(visual=pseudo_rgb, filename=os.path.join(params.debug_outdir, str(params.device) + "_fused_pseudo_rgb.png"))
+
+    return fused_array
+
```
Comment (nfahlgren) [inline plantcv/plantcv/image_fusion.py:79]:
```suggestion
```

Only one blank line at the end of a file

File: tests/tests.py [hunk 10]
```diff
@@ -2259,6 +2259,24 @@ def test_plantcv_image_add():
     added_img = pcv.image_add(gray_img1=img1, gray_img2=img2)
     assert all([i == j] for i, j in zip(np.shape(added_img), TEST_BINARY_DIM))
```
Comment (nfahlgren) [inline tests/tests.py:2261]:
```suggestion


```

Two line returns between functions

File: tests/tests.py [hunk 11]
```diff
@@ -2259,6 +2259,24 @@ def test_plantcv_image_add():
...
 
+def test_plantcv_image_fusion():
```
Comment (nfahlgren) [inline tests/tests.py:2262]: Split into two test functions to test the assertion and RuntimeError separately

File: tests/tests.py [hunk 12]
```diff
@@ -2279,7 +2297,6 @@ def test_plantcv_image_subtract():
     new_img = pcv.image_subtract(img1, img2)
     assert np.array_equal(new_img, np.zeros(np.shape(new_img), np.uint8))
 
-
```
Comment (nfahlgren) [inline tests/tests.py:2282]: Re-add line return to maintain two lines between functions

File: plantcv/plantcv/image_fusion.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    """ Fuse two images of the same size together with given wavelengths representing and make a Spectral_data instance
+    img1: 1st image to be fused
+    img2: 2nd image to be fused
+    wvs1: list of wavelengths represent bands in img1
+    wvs2: list of wavelengths represent bands in img2
+    array_type: (optional) description of the fused array
+    filename: (optional) desired filename of the fused array
+
+    :param img1: np.ndarray
+    :param img2: np.ndarray
```
Comment (nfahlgren) [inline plantcv/plantcv/image_fusion.py:22]:
```suggestion
    :param img1: numpy.ndarray
    :param img2: numpy.ndarray
```

File: plantcv/plantcv/hyperspectral/read_data.py [hunk 14]
```diff
@@ -184,9 +183,6 @@ def read_data(filename):
     pseudo_rgb = _make_pseudo_rgb(spectral_array)
     spectral_array.pseudo_rgb = pseudo_rgb
 
-    if params.debug == "plot":
-        plot_image(pseudo_rgb)
-    elif params.debug == "print":
-        print_image(pseudo_rgb, os.path.join(params.debug_outdir, str(params.device)+"_pseudo_rgb.png"))
+    _debug(visual=pseudo_rgb, filename=os.path.join(params.debug_outdir, str(params.device)+"_pseudo_rgb.png"))
```
Comment (nfahlgren) [inline plantcv/plantcv/hyperspectral/read_data.py:186]:
```suggestion
    _debug(visual=pseudo_rgb, filename=os.path.join(params.debug_outdir, str(params.device) + "_pseudo_rgb.png"))
```

Add spaces between the strings and + operator

File: plantcv/plantcv/__init__.py [hunk 15]
```diff
@@ -94,8 +95,8 @@
...
+           'laplace_filter', 'sobel_filter', 'scharr_filter', 'hist_equalization', 'erode',
+           'image_add', 'image_fusion', 'image_subtract', 'dilate', 'watershed', 'rectangle_mask', 'rgb2gray_hsv', 'rgb2gray_lab',
```
Comment (nfahlgren) [inline plantcv/plantcv/__init__.py:99]: Move `rgb2gray_lab` to a new line and redistribute list to keep the line length < 120 characters

**Review by nfahlgren — CHANGES_REQUESTED** [review]
Hi @DannieSheng, I did a quick check for now just to look at PEP8 formatting issues.

File: plantcv/plantcv/__init__.py (hunk 1 above)
Comment (DannieSheng) [reply plantcv/plantcv/__init__.py:98]: Thank you! I've added it back

**Review by nfahlgren — APPROVED** [review]
Thanks @DannieSheng!

**Pull request merged**